module/zfs/abd.c

```diff
@@ -336,18 +324,52 @@ abd_alloc_pages(abd_t *abd, size_t size)
 		list_del(&page->lru);
 	}
 
-	if (chunks > 1) {
+	/*
+	 * These conditions ensure that a possible transformation to a linear
+	 * ABD would be valid.
+	 */
+	ASSERT(!PageHighMem(sg_page(ABD_SCATTER(abd).abd_sgl)));
```

Comment:
@ahrens this ASSERT here is the root cause of the CentOS 6 failures since `ABD_SCATTER(abd).abd_sgl` hasn't been set yet.  My best guess is this wasn't an issue on other platforms since newer versions of gcc recognized that `PageHighMem()` will always be `0` when `!CONFIG_HIGHMEM` and the entire thing is optimized away.  The following change resolved the issue.

```c
diff --git a/module/zfs/abd.c b/module/zfs/abd.c
index 1f1ab91..d135933 100644
--- a/module/zfs/abd.c
+++ b/module/zfs/abd.c
@@ -324,13 +324,6 @@ abd_alloc_pages(abd_t *abd, size_t size)
                list_del(&page->lru);
        }
 
-       /*
-        * These conditions ensure that a possible transformation to a linear
-        * ABD would be valid.
-        */
-       ASSERT(!PageHighMem(sg_page(ABD_SCATTER(abd).abd_sgl)));
-       ASSERT0(ABD_SCATTER(abd).abd_offset);
-
        if (table.nents == 1) {
                /*
                 * Since there is only one entry, this ABD can be represented
@@ -352,7 +345,13 @@ abd_alloc_pages(abd_t *abd, size_t size)
                 * pages are always mapped into the kernel's address space.
                 * This is not the case for highmem pages, so the
                 * optimization can not be made there.
+                *
+                * These conditions ensure that a possible transformation
+                * to a linear ABD would be valid.
                 */
+               ASSERT(!PageHighMem(sg_page(table.sgl)));
+               ASSERT0(ABD_SCATTER(abd).abd_offset);
+
                abd->abd_flags |= ABD_FLAG_LINEAR;
                abd->abd_flags |= ABD_FLAG_LINEAR_PAGE;
                abd->abd_u.abd_linear.abd_sgl = table.sgl;
```

Reply:
Great catch, thanks!  This was the result of some incomplete code cleanup late in the development cycle :(

I'd prefer to keep the assertion outside the `if` so that it can apply even if we don't actually do the transformation on this call, so I'll fix it to use `table.sgl` instead.